// @threlte/core/dist/index.js
// canvas component
export { default as Canvas } from './Canvas.svelte';
// trait components
export { default as HierarchicalObject } from './internal/HierarchicalObject.svelte';
export { default as SceneGraphObject } from './internal/SceneGraphObject.svelte';
export { default as DisposableObject } from './internal/DisposableObject.svelte';
// components (v6)
export { T, extend } from './components/T/T';
// plugins
export { injectPlugin } from './plugins/injectPlugin';
export { createPlugin } from './plugins/createPlugin';
// hooks
export { useTask } from './hooks/useTask';
export { useStage } from './hooks/useStage';
export { useThrelte } from './hooks/useThrelte';
export { useThrelteUserContext } from './hooks/useThrelteUserContext';
// legacy hooks
// TODO: remove in Threlte 7
export { useFrame } from './hooks/legacy/useFrame';
export { useRender } from './hooks/legacy/useRender';
// useLoader
export { useLoader } from './hooks/useLoader';
// useParent
export { useParent } from './hooks/useParent';
// AsyncWritable
export { asyncWritable } from './lib/asyncWritable';
// utils
export { createObjectStore } from './lib/createObjectStore';
export { createRawEventDispatcher } from './lib/createRawEventDispatcher';
export { watch, memoize, currentWritable } from './lib/storeUtils';
export { forwardEventHandlers } from './lib/forwardEventHandlers';
export { useCache } from './lib/cache';
export { resolvePropertyPath } from './components/T/utils/resolvePropertyPath';
